### ISSUE TITLE
loss function of yolo v3

### ISSUE THREAD

**Issue by mayifeng** [AlexeyAB/darknet#2051]
hi, could u show me where i can find the loss function of yolov3? already found delta for box\class\objectness in yolo_layer.c. but i wonder the code that combine them together. i want to study it and try if i can customize a loss function.

**Comment by mayifeng** [comment]
thank u
can u give me any suggestion on customize a loss function if i want to focus on one delta?
for example, so far the model can get the box correctly but it`s no that correct to tell the class, is it possible (and how) to improve this by customizing a loss function?

**Comment by AlexeyAB** [comment]
Do you want to give 10x advantage to the `objectness` and bounded box `coordinates` over the `probability` of classes?

* change this line: https://github.com/AlexeyAB/darknet/blob/2c04b6ab9a498c99c86d860fc2bf95be1c0fca91/src/yolo_layer.c#L116
to this
`delta[index + stride*class_id] = (1 - output[index + stride*class_id]) / 10; `

* and change this line: https://github.com/AlexeyAB/darknet/blob/2c04b6ab9a498c99c86d860fc2bf95be1c0fca91/src/yolo_layer.c#L143
to this:
`delta[index + stride*n] = (((n == class_id) ? 1 : 0) - output[index + stride*n]) / 10; `

**Comment by muye5** [comment]
Hi, what's the meaning about this chip:
 [darknet/src/yolo_layer.c at Line116](https://github.com/AlexeyAB/darknet/blob/2c04b6ab9a498c99c86d860fc2bf95be1c0fca91/src/yolo_layer.c#L116)
```
    if (delta[index + stride*class_id]){
        delta[index + stride*class_id] = 1 - output[index + stride*class_id];
        if(avg_cat) *avg_cat += output[index + stride*class_id];
        return;
    }
```
I think the class loss is just sum over the 80 classes as bellow
```
        for (n = 0; n < classes; ++n) {
            delta[index + stride*n] = ((n == class_id) ? 1 : 0) - output[index + stride*n];
            if (n == class_id && avg_cat) *avg_cat += output[index + stride*n];
        }
```
can you give a help, thank you!

**Comment by AlexeyAB** [comment]
@muye5 Hi,

> I think the class loss is just sum over the 80 classes as bellow

Yes.

And this code is used for multi-label training:
```cpp
    if (delta[index + stride*class_id]){
        delta[index + stride*class_id] = 1 - output[index + stride*class_id];
        if(avg_cat) *avg_cat += output[index + stride*class_id];
        return;
    }
```

**Comment by muye5** [comment]
> @muye5 Hi,
> 
> > I think the class loss is just sum over the 80 classes as bellow
> 
> Yes.
> 
> And this code is used for multi-label training:
> 
> ```c++
>     if (delta[index + stride*class_id]){
>         delta[index + stride*class_id] = 1 - output[index + stride*class_id];
>         if(avg_cat) *avg_cat += output[index + stride*class_id];
>         return;
>     }
> ```

great, thanks a lot!
